Prevent accidental enablement of efficient attention.

xinyazhang opened the pull request:
Currently Efficient attention and Flash attention share the same set of GPU
kernels on ROCM and have common limitations on head sizes.

Fixes #132004


cc @XilunWu @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o @mcarilli @ptrblck @leslie-fang-intel @jgong5 @voznesenskym @penguinwu @EikanWang @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @LucasLLC @MeetVadakkanchery @mhorowitz @pradeepfn

xinyazhang committed:
Prevent accidental enablement of efficient attention.

This fixes #132004

jithunnair-amd commented:
@xinyazhang Can you please confirm whether the `test_transformers.py::TestSDPAFailureModesCUDA::test_invalid_fused_inputs_head_dim_kernel1_cuda` failure in ROCm CI job for this PR is unrelated to this PR?

xinyazhang commented:
> Can you please confirm whether the `test_transformers.py::TestSDPAFailureModesCUDA::test_invalid_fused_inputs_head_dim_kernel1_cuda` failure in ROCm CI job for this PR is unrelated to this PR?

It seems the UT's have different assumptions on Efficient attention and we need to change the UT accordingly.
I'll go through `test_transformer.py` locally and label keep-going to ensure we didn't miss anything.

xinyazhang committed:
Fix test_invalid_fused_inputs_head_dim_kernel1_cuda and test_unaligned_tensors_cuda

ROCm's Efficient Attention (GPU kernel shared with FA) is more
tolerancing about the inputs.

malfet commented on test/test_transformers.py:
```diff
@@ -1493,6 +1495,8 @@ def test_invalid_fused_inputs_attn_mask_present(self, device, kernel: SDPBackend
     @onlyCUDA
     @unittest.skipIf(not PLATFORM_SUPPORTS_FLASH_ATTENTION, "Does not support fused SDPA or pre-SM80 hardware")
     def test_unaligned_tensors(self, device):
+        if TEST_WITH_ROCM:  # ROCm GPU kernels does not have strict alignment requirements
+            return
```
This comment seems contradictory to what it actually does, feels like you want to skip test with ROCM
```suggestion
```

xinyazhang replied:
This UT acts in an inverted manner: it expects a RuntimeError thrown for irregular shapes to pass, otherwise fails.

However ROCm's implementation does not have such restrictions and no RuntimeError will be thrown, and thus the UT failed because the inputs are acceptable on ROCm.

malfet commented on test/test_transformers.py:
```diff
@@ -1493,6 +1495,8 @@ def test_invalid_fused_inputs_attn_mask_present(self, device, kernel: SDPBackend
     @onlyCUDA
     @unittest.skipIf(not PLATFORM_SUPPORTS_FLASH_ATTENTION, "Does not support fused SDPA or pre-SM80 hardware")
```
```suggestion
    @unittest.skipIf(not PLATFORM_SUPPORTS_FLASH_ATTENTION, "Does not support fused SDPA or pre-SM80 hardware")
    @unittest.skipIf(TEST_WITH_ROCM, "ROCM GPU kernels have strict alignment requirement")
```

malfet reviewed (approved):
LGTM, though marked `test_unaligned_tensors` as skipped, let me know if you disagree

malfet reviewed (changes requested):
This should go into main, shouldn't it? Tried to auto-rebase but it did not work
@jithunnair-amd can you please clarify why PR targets 2.4 rather than main? Were there a more generic change in trunk already?

xinyazhang commented:
> This should go into main, shouldn't it? Tried to auto-rebase but it did not work

Moved the target to main

xinyazhang commented:
Re-creating the PR
